test/framework/machinedeployment_helpers.go

```diff
@@ -137,3 +141,137 @@ func DiscoveryAndWaitForMachineDeployments(ctx context.Context, input DiscoveryA
+	fmt.Fprintf(GinkgoWriter, "Waiting for MachineDeployment rolling upgrade to complete\n")
+	Eventually(func() bool {
+		md := &clusterv1.MachineDeployment{}
+		Expect(input.Getter.Get(ctx, client.ObjectKey{Namespace: input.MachineDeployment.Namespace, Name: input.MachineDeployment.Name}, md)).To(Succeed())
+		return md.Status.Replicas == md.Status.AvailableReplicas
+	}, intervals...).Should(BeTrue())
+}
+
+// UpgradeMachineDeploymentInfrastructureRefAndWaitInput is the input type for UpgradeMachineDeploymentInfrastructureRefAndWait.
+type UpgradeMachineDeploymentInfrastructureRefAndWaitInput struct {
+	ClusterProxy                ClusterProxy
+	Cluster                     *clusterv1.Cluster
+	WaitForMachinesToBeUpgraded []interface{}
+	MachineDeployments          []*clusterv1.MachineDeployment
+}
+
+// UpgradeMachineDeploymentInfrastructureRefAndWait upgrades a machine deployment infrastructure ref and waits for its machines to be upgraded.
+func UpgradeMachineDeploymentInfrastructureRefAndWait(ctx context.Context, input UpgradeMachineDeploymentInfrastructureRefAndWaitInput) {
+	Expect(ctx).NotTo(BeNil(), "ctx is required for UpgradeMachineDeploymentAndWait")
+	Expect(input.ClusterProxy).ToNot(BeNil(), "Invalid argument. input.ClusterProxy can't be nil when calling UpgradeMachineDeploymentAndWait")
+	Expect(input.Cluster).ToNot(BeNil(), "Invalid argument. input.Cluster can't be nil when calling UpgradeMachineDeploymentAndWait")
+	Expect(input.MachineDeployments).ToNot(BeEmpty(), "Invalid argument. input.MachineDeployments can't be empty when calling UpgradeMachineDeploymentAndWait")
+
+	mgmtClient := input.ClusterProxy.GetClient()
+
+	for _, deployment := range input.MachineDeployments {
+		fmt.Fprintf(GinkgoWriter, "Patching the new infrastructure ref to Machine Deployment\n")
+		// Retrieve infra object
+		infraRef := deployment.Spec.Template.Spec.InfrastructureRef
+		infraObj := &unstructured.Unstructured{}
+		infraObj.SetGroupVersionKind(infraRef.GroupVersionKind())
+		key := client.ObjectKey{
+			Namespace: input.Cluster.Namespace,
+			Name:      infraRef.Name,
+		}
+		Expect(mgmtClient.Get(ctx, key, infraObj)).NotTo(HaveOccurred())
+
+		// Creates a new infra object
+		newInfraObj := infraObj
+		newInfraObjName := fmt.Sprintf("%s-%s", infraRef.Name, util.RandomString(6))
+		newInfraObj.SetName(newInfraObjName)
+		newInfraObj.SetResourceVersion("")
+		Expect(mgmtClient.Create(ctx, newInfraObj)).NotTo(HaveOccurred())
+
+		// Patch the new infra object's ref to the machine deployment
+		patchHelper, err := patch.NewHelper(deployment, mgmtClient)
+		Expect(err).ToNot(HaveOccurred())
+		infraRef.Name = newInfraObjName
+		deployment.Spec.Template.Spec.InfrastructureRef = infraRef
+		Expect(patchHelper.Patch(context.TODO(), deployment)).To(Succeed())
+
+		fmt.Fprintf(GinkgoWriter, "Waiting for rolling upgrade to start.\n")
+		WaitForMachineDeploymentRollingUpgradeToStart(ctx, WaitForMachineDeploymentRollingUpgradeToStartInput{
```

Comment:
Is there a particular reason to Wait to start immediately before waiting for complete? 
wondering if we can simplify a little bit and drop the first one

Reply:
This is to understand if rolling upgrade started by just looking at MachineDeployment. Without this check, in cases where upgrade is never initiated, we still see a success because Machine count in MD will be equal to replicas.

Another way to do this check is to iterate over Machines and check for infra ref name. I prefer this way to avoid doing this iteration.

Reply:
Ok, I think I got your point but what about merging this in wait to complete?
In the end, this is a detail of how MD works and having two wait methods in raw is counter-intuitive